### PR TITLE
[Simplifying Deposits UI] Adds the recent deposits table

### LINKED ISSUE #5700: Simplify deposits UI - Add deposit history table showing the last 3 deposits
### Description

This issue is part of the larger Epic to add the Payments Overview Deposits card (https://github.com/Automattic/woocommerce-payments/issues/5686).

As part of this issue, the card should altered to include the following acceptance criteria: 

### Acceptance criteria 

* Add deposit table with the following columns:
    - "Dispatch date"
    - "Status"
    - "Amount"

> **Note**
> Fetching and displaying the data is part of a separate issue. As part of this issue we just need the table to be able to input the data into. 

image

### Designs

P2: paJDYF-7fi-p2

image

### Additional context

Project: pdjTHR-2i5-p2

### PR DESCRIPTION
**Pull request by brucealdridge**
Fixes #5700

#### Changes proposed in this Pull Request

Add a "Deposit History" section showing the three most recent deposits to the new Deposits Overview component.  The top of the Deposits Overview component is shown dimmed in the screenshot below as this has already been implemented in #5799 and is not part of this pull-request.  

Screenshot 2023-03-24 at 2 25 31 PM copy

#### Screenshots

When the Deposits Overview component is loading, recent deposits are not shown in a loading state. A Customer may not have any recent deposits so we either need to show a loading state that then disappears or show no loading state and have content appear.

**Loading Screen**

Screenshot 2023-03-24 at 2 45 28 PM

**Partial Load**
Screenshot 2023-03-24 at 2 45 31 PM

**Fully Loaded**
Screenshot 2023-03-24 at 2 45 34 PM

#### Testing instructions

1. Pull the changes and run `npm run build:deps`
2. Set `_wcpay_feature_simplify_deposits_ui` to `1` in the `wp-options` table
3. From WP Admin go to Payments to view the Overview page
4. Check the Next deposits tables loads as per the designs

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5850` or branch name `issue/5700-recent-deposits-table` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=issue/5700-recent-deposits-table) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c02d66ec23a526d6dab03917d4276953bf74ab3a
- Build time: 2023-04-06 02:52:20 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by shendy-a8c** [comment]
I want to bring an attention to this decision pdjTHR-2jn-p2#comment-3092 that we will stick to the current deposit statuses we have in the deposits list page (/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits).

That means, the UI will look a bit different from one shown on Figma, eg instead of 'Sent' it should say 'Paid'. However, even though on the deposits list page, the status isn't displayed inside a Pill, I think on the Deposits overview card, the status should be displayed inside a Pill.

**Comment by shendy-a8c** [comment]
Another thing I want to bring some attention: I [commented](https://github.com/Automattic/woocommerce-payments/pull/5799#issuecomment-1479079110) that since the row from the recent deposits table and the next deposit table done in https://github.com/Automattic/woocommerce-payments/pull/5799 looks very similar, maybe it's a good idea to make a separate component out of it. Just an idea.

**Comment by brucealdridge** [comment]
> Another thing I want to bring some attention: I [commented](https://github.com/Automattic/woocommerce-payments/pull/5799#issuecomment-1479079110) that since the row from the recent deposits table and the next deposit table done in #5799 looks very similar, maybe it's a good idea to make a separate component out of it. Just an idea.

It's very similar and could be combined. Ad-hoc values are used for `Next Deposit` while the Deposit object is used for recent deposits. It's also worth noting that Next Deposit has a loading state while Recent Deposits does not.

File: client/components/deposits-overview/recent-deposits.tsx
Comment (mattallan) [inline client/components/deposits-overview/recent-deposits.tsx]: We should move this to the `deposits-overview/utils.ts` to keep this file with just the components

Also small nit-pick, I think this function should start with lowercase

File: client/components/deposits-overview/index.tsx
Comment (mattallan) [inline client/components/deposits-overview/index.tsx:28]: Just making a note that this will be updated by @Jinksi PR: https://github.com/Automattic/woocommerce-payments/pull/5822

File: client/components/deposits-overview/recent-deposits.tsx
Comment (mattallan) [inline client/components/deposits-overview/recent-deposits.tsx]: Another note that this will need to be updated to `<DepositStatusPill />` (added https://github.com/Automattic/woocommerce-payments/pull/5946). I don't think we should hold off merging this though.

We can update that other PR to update this status pill after this has been merged.

**Comment by mattallan** [comment]
While testing this PR I noticed the Deposits History loads quicker than the next deposits section. Here's a screencast of the issue:

https://user-images.githubusercontent.com/2275145/229669156-e1f5e071-c014-4c29-b85d-d40173cc0a02.mp4

---

This is only a small issue but I'm wondering if we should be passing the isLoading prop along to `<RecentDeposits />` so that all of the components are loading together or at the very least, in order from top to bottom.

For example:

```
<NextDepositDetails isLoading={ isLoading } overview={ overview } />

<RecentDeposits isLoading={ isLoading } currency={ currency } />
```

And then the condition in `RecentDeposits` would still wait for recentDeposits to load but also check the main component is not still loading, i.e.

```
const recentDeposits = GetRecentDeposits( currency );

if ( isLoading || recentDeposits.isLoading || recentDeposits.deposits.length === 0 ) {
	return <></>;
}
```

**Review by mattallan — COMMENTED** [review]
Nice PR @brucealdridge! I've left a few notes about moving the function into the utils file and tweaking how the the recent deposits table is loading before the rest of the component but apart from that this is looking good!

FWIW I passed your code to chatGPT to see if it had any suggestions and it was pretty minor stuff that I don't think is overly worth providing feedback on 😄

**Comment by haszari** [comment]
> This is only a small issue but I'm wondering if we should be passing the isLoading prop along to <RecentDeposits /> so that all of the components are loading together or at the very least, in order from top to bottom.

That sounds like a micro-optimisation or premature optimisation IMO :) 

Having individual elements load as quick as they can seems shippable, usable and maintainable to me (less coupling). 

If design / product (or engineers!) think this is important we could log an issue and handle as follow up.

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits.tsx]: Nit: this can be moved to "External dependencies"

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits.tsx]: Nit: this should be `RecentDepositsProps` to align with the function definition.

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits.tsx:34]: 🙌  This looks to me like a custom React hook, so it should use the [`use` naming convention](https://react.dev/learn/reusing-logic-with-custom-hooks#hook-names-always-start-with-use).

e.g. `useRecentDeposits`

File: client/components/deposits-overview/test/index.tsx
Comment (Jinksi) [inline client/components/deposits-overview/test/index.tsx]: Note: [`getByText` will fail the test if it can't find an element](https://testing-library.com/docs/queries/about#types-of-queries) (or there is > 1), so `toBeTruthy` is not required.

```suggestion
		getByText( 'January 2, 2020' );
```

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits.tsx]: Thinking out loud: [returning `null` is a more common pattern](https://testing-library.com/docs/queries/about#types-of-queries) than returning an empty fragment.

However, I notice that `<></>` is used throughout WCPay, perhaps to stick to `JSX.Element` return type.

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits.tsx]: Nit: `DepositRowProps` isn't also required here, as it's passed to ` React.FC< DepositRowProps >`

```suggestion
} ): JSX.Element => {
```

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits.tsx]: 🙌  This looks to me like a custom React hook (its calling another hook, `useDeposits`), so it should use the [`use` naming convention](https://react.dev/learn/reusing-logic-with-custom-hooks#hook-names-always-start-with-use).

```suggestion
const useRecentDeposits = ( currency?: string ): RecentDeposits => {
```

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits.tsx]: This component's variable name `RecentDeposits` is the same as the `RecentDeposits` TS interface used for the custom hook above.

I'm not exactly sure what impact this has, but I would consider changing either one to avoid confusion.

File: client/components/deposits-overview/recent-deposits.tsx
Comment (haszari) [reply client/components/deposits-overview/recent-deposits.tsx]: > However, I notice that <></> is used throughout WCPay, perhaps to stick to JSX.Element return type.

Curious, if we continue doing this (using empty fragments instead of null) I want to understand the benefit/rationale.

Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits.tsx]: At a glance, my understanding is that this is mainly to appease TS:
* `<></>` is a valid return for the `JSX.Element` type.
* `null` is not a `JSX.Element`, we'd have to add `JSX.Element | null` to the function return type (IMO good thing, more explicit).

I think both options clearly communicate the intention (return empty/nothing).

<details>
<summary>There is a negligible performance hit to using `<></>` compared to `null`</summary>

In this contrived example, you'll see that `<ReturnFragment />` is rendered, shown in the React profiler as `0.1ms`, while `<ReturnNull />` isn't rendered.

Screenshot 2023-04-05 at 10 23 02

image

</details>

File: client/components/deposits-overview/recent-deposits.tsx
Comment (brucealdridge) [reply client/components/deposits-overview/recent-deposits.tsx]: I thought so too, a linter failed me on it with lowercase, maybe due to the naming?

```
ESLint: React Hook "useDeposits" is called in function "getRecentDeposits" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter.(react-hooks/rules-of-hooks)
```

File: client/components/deposits-overview/recent-deposits.tsx
Comment (brucealdridge) [reply client/components/deposits-overview/recent-deposits.tsx]: By this, I'm assuming you mean just the `Link` rather than the four lines

File: client/components/deposits-overview/recent-deposits.tsx
Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits.tsx]: Nice catch! Interesting that that config missed this when named `Get` rather than `get`. I wonder if this is intentional and it assumed `Get` is a react component function.

image

**Comment by brucealdridge** [comment]
> While testing this PR I noticed the Deposits History loads quicker than the next deposits section. Here's a screencast of the issue:

That's interesting, in my testing the recent deposits were always the slowest to load. I'm curious what others think about the potential loading order

File: client/components/deposits-overview/recent-deposits.tsx
Comment (brucealdridge) [reply client/components/deposits-overview/recent-deposits.tsx]: Thanks, clearly missed this copy/paste

File: client/components/deposits-overview/recent-deposits.tsx
Comment (haszari) [reply client/components/deposits-overview/recent-deposits.tsx]: Gotcha – it's simpler to guarantee `JSX.Element` return type. Works for me :)

File: client/components/deposits-overview/recent-deposits.tsx
Comment (haszari) [reply client/components/deposits-overview/recent-deposits.tsx]: > I thought so too, a linter failed me on it with lowercase, maybe due to the naming?

> ESLint: React Hook "useDeposits" is called in function "getRecentDeposits" that is neither a React function component nor a custom React Hook function. 

I think the linter rule AND @Jinksi are saying the same thing - we need to be clear about whether this is a component, or a custom hook. This looks like a custom hook to me too, so we should follow the linter recommendation and make it a true custom hook.

File: client/components/deposits-overview/recent-deposits-list.tsx
Comment (Jinksi) [inline client/components/deposits-overview/recent-deposits-list.tsx]:
```suggestion
interface RecentDeposits {
```

Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits-list.tsx]: Did you mean to call this one `RecentDeposits`?

File: client/components/deposits-overview/recent-deposits-list.tsx
Comment (mattallan) [inline client/components/deposits-overview/recent-deposits-list.tsx]: Not really 100% certain on this, so maybe @Jinksi can confirm, but should we move this `useRecentDeposits` hook into a `hooks.ts`?
I noticed this is what we did in the account-balances component

File: includes/admin/class-wc-payments-admin.php
Comment (mattallan) [inline includes/admin/class-wc-payments-admin.php]: Assuming this is dummy/test data?

File: client/components/deposits-overview/recent-deposits-list.tsx
Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits-list.tsx]: I don't have a strong reason for or against including in a new `hooks.ts` file, since this hook is only used within this file.

However, after a quick search of WCPay shows, it looks like including hooks in `hooks.ts` is a more common practice.

Comment (haszari) [reply client/components/deposits-overview/recent-deposits-list.tsx]: I think a dedicated, explicit hooks file is good practice. Encourages decoupling and thinking about hook interface in isolation.

File: includes/admin/class-wc-payments-admin.php
Comment (brucealdridge) [reply includes/admin/class-wc-payments-admin.php]: TY, Good catch! The dangers of using `git commit -a`!

File: client/components/deposits-overview/recent-deposits-list.tsx
Comment (Jinksi) [reply client/components/deposits-overview/recent-deposits-list.tsx]: Resolved in 19074f12f756dfb8325eac210f8a4ba6c36f0251

**Review by mattallan — APPROVED** [review]
I've reviewed the latest code changes and ran through a test on local and it looks all good.

Happy to approve this one! 💯

**Pull request merged**
